unc_classification_tab_data/models/vae.py

```diff
@@ -0,0 +1,325 @@
+import torch
+import torch.nn as nn
+import numpy as np
+import torch.distributions as dist
+import torch.utils.data
+from typing import List, Tuple
+from .constants import *
+
+
+class Encoder(nn.Module):
+    """The encoder module, which encodes an input into the latent space.
+
+    Parameters
+    ----------
+    input_dim: int
+        The input dimensionality.
+    hidden_dims: List[int]
+        A list with the sizes of the hidden layers.
+    z_dim: int
+        The size of the latent space.
+    """
+
+    def __init__(self, input_dim: int, hidden_dims: List[int], z_dim: int):
+        super().__init__()
+        architecture = [input_dim] + hidden_dims
+        self.hidden_layers = []
+        if hidden_dims == []:
+            self.hidden_layers = []
+        else:
+            for i in range(len(architecture) - 1):
+                self.hidden_layers.append(nn.Linear(architecture[i], architecture[i + 1]))
+                self.hidden_layers.append(nn.ReLU())
+        self.hidden = nn.Sequential(*self.hidden_layers)
+        self.mean = nn.Linear(architecture[-1], z_dim)
+        self.log_std = nn.Linear(architecture[-1], z_dim)
+        self.std = nn.Sigmoid()
+
+    def forward(self, input_tensor: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
+        """Perform forward pass of encoder. Returns mean and standard deviation corresponding to
+        an independent Normal distribution.
+
+        Parameters
+        ----------
+        input_tensor: torch.Tensor
+            The input to the encoder.
+        """
+        h = self.hidden(input_tensor)
+        mean = self.mean(h)
+        log_std = self.log_std(h)
+        std = self.std(log_std)
+        return mean, std
+
+
+class Decoder(nn.Module):
+    """The decoder module, which decodes a sample from the latent space back to the space of
+    the input data.
+
+    Parameters
+    ----------
+    input_dim: int
+        The dimensionality of the input
+    hidden_dims: List[int]
+        A list with the sizes of the hidden layers.
+    z_dim: int
+        The size of the latent space.
+    """
+
+    def __init__(self, input_dim: int, hidden_dims: List[int], z_dim: int):
+        super().__init__()
+        architecture = [z_dim] + hidden_dims
+        self.hidden_layers = []
+        if hidden_dims == []:
```

Comment:
this is very un-pythonic, and it also constructs an extra list
```suggestion
        if not hidden_dims:
```